data_model/src/executor.rs

```diff
@@ -47,10 +52,87 @@ mod model {
         pub wasm: WasmSmartContract,
     }
 
+    /// Executor data model.
+    ///
+    /// Defined from within the executor, it describes certain structures the executor
+    /// works with.
+    ///
+    /// Executor can define:
+    ///
+    /// - Permission tokens (see [`crate::permission::PermissionToken`])
+    /// - Configuration parameters (see [`crate::parameter::Parameter`])
+    #[derive(
+        Default,
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Constructor,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+        Display,
+    )]
+    #[ffi_type]
+    #[display(fmt = "{self:?}")]
+    pub struct ExecutorDataModel {
+        /// Data model JSON schema, typically produced by [`IntoSchema`].
+        pub schema: JsonString,
+        /// Permission tokens supported by the executor.
+        ///
+        /// These IDs refer to the types in the schema.
+        pub permission_token_ids: BTreeSet<PermissionTokenId>,
+        /// Configuration parameters supported by the executor.
+        ///
+        /// These IDs refer to the types in the schema.
+        pub parameter_ids: BTreeSet<ParameterId>,
+    }
+
     // TODO: Client doesn't need structures defined inside this macro. When dynamic linking is
     // implemented use: #[cfg(any(feature = "transparent_api", feature = "ffi_import"))]
 }
 
+impl ExecutorDataModel {
+    /// Getter
+    pub fn schema(&self) -> &JsonString {
+        &self.schema
+    }
+
+    /// Getter
+    pub fn permission_token_ids(&self) -> &BTreeSet<PermissionTokenId> {
+        &self.permission_token_ids
+    }
+
+    /// Getter
+    pub fn parameter_ids(&self) -> &BTreeSet<ParameterId> {
+        &self.parameter_ids
+    }
+}
+
+/// Defines an object that is part of the [`ExecutorDataModel`].
+///
+/// An object consists of id and payload.
+/// ID is a unique type identifier in the executor's data model.
+/// Payload is JSON-serialized type structure, according to the type schema.
+// TODO: support both JSON and SCALE for payload.
+pub trait ExecutorDataModelObject: Sized {
+    /// Generic identification
+    type DefinitionId: PartialEq + Clone + Into<crate::name::Name>;
+
+    /// Constructor
+    fn new(definition_id: Self::DefinitionId, payload: JsonString) -> Self;
+
+    /// Getter for the object id
+    fn object_definition_id(&self) -> &Self::DefinitionId;
+
+    /// Getter for the object payload
+    fn object_payload(&self) -> &JsonString;
```

Comment:
```suggestion
    fn id(&self) -> &Self::DefinitionId;

    /// Getter for the object payload
    fn schema(&self) -> &JsonString;
```

Reply:
As for `id`, well, let's rename this field from `definition_id` to `id` in `PermissionToken` and `Parameter` structs as well then. **But,** the reason I added `object_` prefix here is because it created ambiguity for Rust, because structs themselves also had `id()` getters in their impls.

But that's not going to be an issue if I move this trait into `iroha_smart_contract`, as you suggested.

As for `schema`, it's not, isn't it? It's a JSON-encoded permission structure, i.e. payload, not the schema of it.

Reply:
> As for schema, it's not, isn't it? It's a JSON-encoded permission structure, i.e. payload, not the schema of it.

true, then `payload` instead of `object_payload`

Reply:
Done